packages/voting-contract/src/lib.rs

```diff
@@ -197,20 +199,34 @@ where
 {
     // anyone can trigger this if the vote passed
 
-    let mut prop = proposals::<P>().load(deps.storage, proposal_id)?;
-    if [Status::Executed, Status::Rejected, Status::Passed]
+    let mut prop = proposals().load(deps.storage, proposal_id)?;
+    let previous_status = prop.status;
+    prop.update_status(&env.block);
+    proposals::<P>().save(deps.storage, proposal_id, &prop)?;
+
+    if [Status::Executed, Status::Passed]
         .iter()
         .any(|x| *x == prop.status)
     {
         return Err(ContractError::WrongCloseStatus {});
     }
-    if !prop.expires.is_expired(&env.block) {
+    if prop.status == Status::Open {
         return Err(ContractError::NotExpired {});
     }
 
-    // set it to failed
-    prop.status = Status::Rejected;
-    proposals::<P>().save(deps.storage, proposal_id, &prop)?;
+    // Two scenarios:
+    // 1) any status -> rejected (update_status changed it)
+    // 2) proposal was already rejected (double close)
+    // To differentiate those, we need to compare current status with previous one
+    // If that condition is false, then it is first time closing this proposal
+    if prop.status == Status::Rejected && previous_status == Status::Rejected {
```

Comment:
I think you might as well just check the status at the beginning of the fn and return `ContractError::Rejected` if the status was already `Rejected`, then update it, and so on. Should end up with simpler logic.

Reply:
Agree.